htlcswitch/mailbox_test.go

```diff
@@ -315,6 +318,106 @@ func TestMailBoxFailAdd(t *testing.T) {
 	// duplicate fails are sent.
 	go failAdds(adds)
 	ctx.checkFails(nil)
+
+}
+
+// TestMailBoxPacketPrioritization asserts that the mailbox will prioritize
+// delivering Settle and Fail packets over Adds if both are available for
+// delivery at the same time.
+func TestMailBoxPacketPrioritization(t *testing.T) {
+	t.Parallel()
+
+	// First, we'll create new instance of the current default mailbox
+	// type.
+	mailBox := newMemoryMailBox(&mailBoxConfig{
+		clock:  clock.NewDefaultClock(),
+		expiry: time.Minute,
+	})
+	mailBox.Start()
+	defer mailBox.Stop()
+
+	const numPackets = 5
+
+	_, _, aliceChanID, bobChanID := genIDs()
+
+	// Next we'll send the following sequence of packets:
+	//  - Settle1
+	//  - Add1
+	//  - Add2
+	//  - Fail
+	//  - Settle2
+	sentPackets := make([]*htlcPacket, numPackets)
+	for i := 0; i < numPackets; i++ {
+		pkt := &htlcPacket{
+			outgoingChanID: aliceChanID,
+			outgoingHTLCID: uint64(i),
+			incomingChanID: bobChanID,
+			incomingHTLCID: uint64(i),
+			amount:         lnwire.MilliSatoshi(prand.Int63()),
+		}
+
+		switch i {
+		case 0, 4:
+			// First and last packets are a Settle. A non-Add is
+			// sent first to make the test deterministic w/o needing
+			// to sleep.
+			pkt.htlc = &lnwire.UpdateFulfillHTLC{ID: uint64(i)}
+		case 1, 2:
+			// Next two packets are Adds.
+			pkt.htlc = &lnwire.UpdateAddHTLC{ID: uint64(i)}
+		case 3:
+			// Last packet is a Fail.
+			pkt.htlc = &lnwire.UpdateFailHTLC{ID: uint64(i)}
+		}
+
+		sentPackets[i] = pkt
+
+		err := mailBox.AddPacket(pkt)
+		if err != nil {
+			t.Fatalf("failed to add packet: %v", err)
+		}
+	}
+
```

Comment:
How stable is this test? Seems possible that the two adds could go out first if the consumer acts fast enough. Other than that, seems like a solid test.

Reply:
that's a good q, it's totally deterministic. the first packet added is a reply, so it's guaranteed to be delivered first. additionally all packets are added before we start consuming, so by the second iteration the courier is guaranteed to be aware of all other reply packets that should be prioritized over the adds